### PR TITLE
issue #1223 a port of Mime.processDecoded()

### LINKED ISSUE #1223: a port of Mime.processDecoded (to MsgBlocks)
For decoding mime content, in TS we use this method:

```ts
  public static decode = async (mimeMsg: Uint8Array): Promise<MimeContent> => {
    let mimeContent: MimeContent = { attachments: [], headers: {}, subject: undefined, text: undefined, html: undefined, signature: undefined, from: undefined, to: [], cc: [], bcc: [] };
    const parser = new MimeParser();
    const leafNodes: { [key: string]: MimeParserNode } = {};
    parser.onbody = (node: MimeParserNode) => {
      const path = String(node.path.join('.'));
      if (typeof leafNodes[path] === 'undefined') {
        leafNodes[path] = node;
      }
    };
    return await new Promise((resolve, reject) => {
      try {
        parser.onend = () => {
          try {
            for (const name of Object.keys(parser.node.headers)) {
              mimeContent.headers[name] = parser.node.headers[name][0].value;
            }
            mimeContent.rawSignedContent = Mime.retrieveRawSignedContent([parser.node]);
            for (const node of Object.values(leafNodes)) {
              if (Mime.getNodeType(node) === 'application/pgp-signature') {
                mimeContent.signature = node.rawContent;
              } else if (Mime.getNodeType(node) === 'text/html' && !Mime.getNodeFilename(node)) {
                // html content may be broken up into smaller pieces by attachments in between
                // AppleMail does this with inline attachments
                mimeContent.html = (mimeContent.html || '') + Mime.getNodeContentAsUtfStr(node);
              } else if (Mime.getNodeType(node) === 'text/plain' && (!Mime.getNodeFilename(node) || Mime.isNodeInline(node))) {
                mimeContent.text = (mimeContent.text ? `${mimeContent.text}\n\n` : '') + Mime.getNodeContentAsUtfStr(node);
              } else if (Mime.getNodeType(node) === 'text/rfc822-headers') {
                if (node._parentNode && node._parentNode.headers.subject) {
                  mimeContent.subject = node._parentNode.headers.subject[0].value;
                }
              } else {
                mimeContent.attachments.push(Mime.getNodeAsAttachment(node));
              }
            }
            const headers = Mime.headerGetAddress(mimeContent, ['from', 'to', 'cc', 'bcc']);
            mimeContent.subject = String(mimeContent.subject || mimeContent.headers.subject || '');
            mimeContent = Object.assign(mimeContent, headers);
            resolve(mimeContent);
          } catch (e) {
            reject(e);
          }
        };
        parser.write(mimeMsg);
        parser.end();
      } catch (e) { // todo - on Android we may want to fail when this happens, evaluate effect on browser extension
        Catch.reportErr(e);
        resolve(mimeContent);
      }
    });
  }
```

But @DenBond7 probably already has some method for decoding mime. Therefore I don't want to reimplement the above, but instead use Den's methods (and its outputs) for the above purpose (if possible)

Then you take outputs from the MIME decode, and you feed them into a new method that needs implementing, similar to the one below:

```ts
  public static processDecoded = (decoded: MimeContent): MimeProccesedMsg => {
    const blocks: MsgBlock[] = [];
    if (decoded.text) {
      const blocksFromTextPart = MsgBlockParser.detectBlocks(Str.normalize(decoded.text)).blocks;
      // if there are some encryption-related blocks found in the text section, which we can use, and not look at the html section
      if (blocksFromTextPart.find(b => b.type === 'encryptedMsg' || b.type === 'signedMsg' || b.type === 'publicKey' || b.type === 'privateKey')) {
        blocks.push(...blocksFromTextPart); // because the html most likely containt the same thing, just harder to parse pgp sections cause it's html
      } else if (decoded.html) { // if no pgp blocks found in text part and there is html part, prefer html
        blocks.push(MsgBlock.fromContent('plainHtml', decoded.html));
      } else { // else if no html and just a plain text message, use that
        blocks.push(...blocksFromTextPart);
      }
    } else if (decoded.html) {
      blocks.push(MsgBlock.fromContent('plainHtml', decoded.html));
    }
    for (const file of decoded.attachments) {
      const treatAs = file.treatAs();
      if (treatAs === 'encryptedMsg') {
        const armored = PgpArmor.clip(file.getData().toUtfStr());
        if (armored) {
          blocks.push(MsgBlock.fromContent('encryptedMsg', armored));
        }
      } else if (treatAs === 'signature') {
        decoded.signature = decoded.signature || file.getData().toUtfStr();
      } else if (treatAs === 'publicKey') {
        blocks.push(...MsgBlockParser.detectBlocks(file.getData().toUtfStr()).blocks);
      } else if (treatAs === 'privateKey') {
        blocks.push(...MsgBlockParser.detectBlocks(file.getData().toUtfStr()).blocks);
      } else if (treatAs === 'encryptedFile') {
        blocks.push(MsgBlock.fromAttachment('encryptedAttachment', '', { name: file.name, type: file.type, length: file.getData().length, data: file.getData() }));
      } else if (treatAs === 'plainFile') {
        blocks.push(MsgBlock.fromAttachment('plainAttachment', '', {
          name: file.name, type: file.type, length: file.getData().length, data: file.getData(), inline: file.inline, cid: file.cid
        }));
      }
    }
    if (decoded.signature) {
      for (const block of blocks) {
        if (block.type === 'plainText') {
          block.type = 'signedText';
          block.signature = decoded.signature;
        } else if (block.type === 'plainHtml') {
          block.type = 'signedHtml';
          block.signature = decoded.signature;
        }
      }
      if (!blocks.find(block => ['plainText', 'plainHtml', 'signedMsg', 'signedHtml', 'signedText'].includes(block.type))) { // signed an empty message
        blocks.push(new MsgBlock("signedMsg", "", true, decoded.signature));
      }
    }
    return { headers: decoded.headers, blocks, from: decoded.from, to: decoded.to, rawSignedContent: decoded.rawSignedContent };
  }
```

### PR DESCRIPTION
**Pull request by IvanPizhenko**
This PR introduces port of the function `Mime.processDecoded()` from Typescript to Kotlin. 

close #1223 // if this PR closes an issue

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities

### CODE REVIEW

**Review by tomholub — COMMENTED** [review]
This looks good to me overall (didn't dig into details) :+1:

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/response/model/node/EncryptedAttMsgBlock.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import android.net.Uri
+import android.os.Parcel
+import android.os.Parcelable
+import com.google.gson.annotations.Expose
+
+data class EncryptedAttMsgBlock(@Expose override val content: String?,
+                                @Expose val attMeta: AttMeta) : MsgBlock {
+
+  var fileUri: Uri? = null
```
[the remark this answers is not in the document]
Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/response/model/node/EncryptedAttMsgBlock.kt:17]: Yup. It's important for performance.

**Comment by DenBond7** [comment]
@IvanPizhenko I had to add some changes to fix the JUnit tests that relate to Android. Please look at them.

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/response/model/node/MsgBlock.kt [hunk 2]
```diff
@@ -86,13 +93,28 @@ interface MsgBlock : Parcelable {
         override fun newArray(size: Int): Array<Type?> = arrayOfNulls(size)
       }
 
+      @JvmField
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/response/model/node/MsgBlock.kt:96]: I think using `@JvmField` is redundant over the whole codebase(except a case when we really need that, for example, `Parcelable` implementation).

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt [hunk 3]
```diff
@@ -256,6 +268,324 @@ object PgpMsg {
...
+    )
+  }
+
+  // Typescript:  public static processDecoded = (decoded: MimeContent): MimeProcessedMsg
+  fun processDecodedMimeMessage(decoded: MimeContent): MimeProcessedMsg {
+    val blocks = mutableListOf<MsgBlock>()
+    if (decoded.text != null) {
+      val blocksFromTextPart = MsgBlockParser.detectBlocks(decoded.text)
+      val suitableBlock = blocksFromTextPart.firstOrNull {
+        it.type == MsgBlock.Type.ENCRYPTED_MSG || it.type == MsgBlock.Type.SIGNED_MSG
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:377]: Can it be simplified with `it.type in MsgBlock.Type.wellKnownBlockTypes`?

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt [hunk 4]
```diff
@@ -256,6 +268,324 @@ object PgpMsg {
...
+      text = text?.toString(),
+      from = msg.from,
+      to = msg.getRecipients(Message.RecipientType.TO),
+      cc = msg.getRecipients(Message.RecipientType.CC),
+      bcc = msg.getRecipients(Message.RecipientType.BCC)
+    )
+  }
+
+  // Typescript:  public static processDecoded = (decoded: MimeContent): MimeProcessedMsg
+  fun processDecodedMimeMessage(decoded: MimeContent): MimeProcessedMsg {
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:372]: This method is too long. Can we chunk it?

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt [hunk 5]
```diff
@@ -256,6 +268,324 @@ object PgpMsg {
...
+    "Message.pgp", "openpgp-encrypted-message.asc"
+  )
+
+  private val encryptedFileRegex = Regex("(\\.pgp\$)|(\\.gpg\$)|(\\.[a-zA-Z0-9]{3,4}\\.asc\$)")
+  private val privateKeyRegex = Regex("(cryptup|flowcrypt)-backup-[a-z0-9]+\\.(key|asc)\$")
+  private val publicKeyRegex1 = Regex("^(0|0x)?[A-F0-9]{8}([A-F0-9]{8})?.*\\.asc\$")
+  private val publicKeyRegex2 = Regex("[A-F0-9]{8}.*\\.asc\$")
 }
+
 //                   || ex.message?.contains("Invalid Curve25519 public key") == true
```
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:590]: Please remove it or describe why we should leave it as-is

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt [hunk 6]
```diff
@@ -265,4 +282,80 @@ class PgpMsgTest {
...
+      val expectedComplete = expectedBlock["complete"].asBoolean
+      val actualBlock = processed.blocks[i]
+      val actualContent = (actualBlock.content ?: "").normalizeEol().removeUtf8Bom()
+
+      assertEquals(expectedBlockType, actualBlock.type)
+      assertEquals(expectedComplete, actualBlock.complete)
+      assertEquals(expectedContent, actualContent)
+
+      if (
+        actualBlock.type == MsgBlock.Type.SIGNED_TEXT
```
Comment (DenBond7) [inline FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:352]: can be simplified to one line 

`if (actualBlock.type in setOf(MsgBlock.Type.SIGNED_TEXT, MsgBlock.Type.SIGNED_HTML)) `

**Review by DenBond7 — CHANGES_REQUESTED** [review]
Please look at the comments

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/javax/mail/PartExt.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+/*
+ * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
+ * Contributors: ivan
```
[the remark this answers is not in the document]
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/extensions/javax/mail/PartExt.kt:3]: I assume new file template puts my Linux user name here. I usually change it to my full name, but I missed this one.

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt (hunk 6 above)
Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:352]: ok

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt (hunk 3 above)
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:377]: Yes, I think so, didn't notice this.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt (hunk 5 above)
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:590]: will remove this

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt (hunk 4 above)
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:372]: I think we can, but not a lot. will try to do that

**Comment by IvanPizhenko** [comment]
@DenBond7 I've made some fixes, please recheck it

**Review by DenBond7 — APPROVED** [review]

**Pull request merged**